### PR TITLE
Make congruence poset object a digraph

### PR DESCRIPTION
**Pull request by mtorpey**
This probably should've been the case when the object was first created.  A poset of semigroup congruences is a digraph with vertices labelled by the congruences themselves.  An edge `i -> j` exists if congruence ~~`i`~~ `j` is a subrelation (a refinement) of congruence ~~`j`~~ `i`.

This is in reference to Issue #369.

**EDIT:** `i` and `j` reversed.

### CODE REVIEW

**Comment by wilfwilson** [comment]
> An edge `i -> j` exists if congruence `i` is a subrelation (a refinement) of congruence `j`.

That's not what you've implemented:
```
gap> S := PartitionMonoid(2);;
gap> lat := LatticeOfCongruences(S);;
gap> univ := UniversalSemigroupCongruence(S);;
gap> triv := SemigroupCongruence(S, []);;
gap> pos_univ := Position(DigraphVertexLabels(lat), univ);;
gap> pos_triv := Position(DigraphVertexLabels(lat), triv);;
gap> IsDigraphEdge(lat, [pos_univ, pos_triv]);
true
gap> IsSubrelation(triv, univ);
false
```
Which do you want?

**Comment by mtorpey** [comment]
Where does that quote come from?  My documentation for `IsCongruencePoset` says

> an edge from vertex `i`  to  vertex  `j` if and only if the congruence numbered `j` is a subrelation of the congruence numbered `i`.

which is what I intend.

**Comment by wilfwilson** [comment]
Thanks for clearing that up @mtorpey. My comments in this post are all about making sure that the behaviour of your posets is consistent with Digraphs.

Here's the documentation:

> A congruence poset is a digraph with a vertex for each congruence, and an edge from vertex `i` to vertex `j` if and only if the congruence numbered `j` is a subrelation of the congruence numbered `i`.

You should make your posets reflexive, since a congruence contains itself. Furthermore, it makes sense to do this for the following reason. Your posets are currently antisymmetric and transitive, but since they're not reflexive, `IsPartialOrderDigraph` returns `false`:

```
gap> S := SymmetricInverseMonoid(3);;
gap> lat := LatticeOfCongruences(S);;
gap> IsAntisymmetricDigraph(lat) and IsTransitiveDigraph(lat);
true
gap> IsReflexiveDigraph(lat) or IsPartialOrderDigraph(lat);
false
```

I think it's natural to expect a poset of congruences to satisfy `IsPartialOrderDigraph`. If you make this change, then when a poset of congruences is created, you can set `IsPartialOrderDigraph` to be `true` (or you can make this inherent for any poset of congruences).

More fundamentally, you should change the definition of the partial order that is used for your posets. In your setup, `x -> y` is an edge if `y ⊆ x`. However this doesn't work well with Digraphs. From the Digraphs documentation:

> For a partial order digraph `digraph`, the corresponding partial order is the relation `≤`, defined by `x≤y` if and only if `[x, y]` (aka `x -> y`) is an edge of digraph.

So, taking this all together, `cong1 ≤ cong2` if and only if `cong2 ⊆ cong1`, which I hope you can see is strange. Here's an example of where the problem kicks in:

The *join* of two elements partially ordered by less-than-or-equal is their least unique upper bound. Therefore, with your congruences, the join of two elements `c1` and `c2` is greater than both `c1` and `c2`, and is therefore *contained* in `c1` and `c2`. But (obviously) the join of `c1` and`c2` *contains* `c1` and `c2`. This means that, according to Digraphs, a `JoinSemilatticeOfCongruences` is a meet semilattice rather than a join semilattice.

```
gap> S := Semigroup([
>  Transformation([1, 3, 3, 3]),
>  Transformation([1, 2, 1])]);;
gap> cong1 := SemigroupCongruence(S,
>  [Transformation([1, 1, 1, 1]), Transformation([1, 2, 1])]);;
gap> cong2 := SemigroupCongruence(S,
>  [Transformation([1, 1, 1, 1]), Transformation([1, 3, 3, 3])]);;
gap> poset := JoinSemilatticeOfCongruences([cong1, cong2],
>   JoinSemigroupCongruences);;
gap> poset := DigraphAddAllLoops(poset);; # make it a partial order
gap> IsPartialOrderDigraph(poset);
true
gap> IsJoinSemilatticeDigraph(poset);
false
gap> IsMeetSemilatticeDigraph(poset);
true
```

Let me know if you want more convincing about this.

File: gap/congruences/conglatt.gi [hunk 1]
```diff
@@ -125,19 +115,19 @@ SEMIGROUPS.PrincipalXCongruencePosetNC :=
...
   SetUnderlyingSemigroupOfCongruencePoset(poset, S);
+  SetFilterObj(poset, IsCongruencePoset);
   return poset;
 end;
 
 InstallMethod(MinimalCongruences,
 "for a congruence poset",
 [IsCongruencePoset],
-poset -> CongruencesOfPoset(poset){DigraphSinks(poset!.po)});
+poset -> CongruencesOfPoset(poset){DigraphSinks(poset)});
```
Comment (wilfwilson) [inline gap/congruences/conglatt.gi]: When you add loops at every vertex, you'll have to change this. Options: vertices with out-degree 1, or vertices with in-degree 1. Or remove loops and then get the sinks.

File: doc/conglatt.xml [hunk 2]
```diff
@@ -21,11 +21,12 @@
...
       congruences in a congruence poset can be left, right, or two-sided. <P/>
 
-      A congruence poset is displayed as a list of lists, which describes the
-      partial order of its congruences: the integer <C>j</C> appears in list
-      <C>i</C> if and only if the congruence numbered <C>j</C> is a subrelation
-      of the congruence numbered <C>i</C>.  The list of congruences can be
-      obtained using <Ref Attr = "CongruencesOfPoset"/>.
+      A congruence poset is a digraph
+      (see <Ref Filt = "IsDigraph" BookName = "Digraphs"/>) with a vertex for
+      each congruence, and an edge from vertex <C>i</C> to vertex <C>j</C> if
```
Comment (wilfwilson) [inline doc/conglatt.xml]: Swap this

File: doc/conglatt.xml [hunk 3]
```diff
@@ -35,20 +36,21 @@
...
-[ [  ], [ 1 ], [ 1, 2, 4 ], [ 1, 2 ] ]
+<poset of 4 congruences over <symmetric inverse monoid of degree 2>>
 gap> IsCongruencePoset(poset);
 true
-gap> poset[3];
-[ 1, 2, 4 ]
+gap> IsDigraph(poset);
+true
+gap> OutNeighbours(poset);
+[ [  ], [ 1 ], [ 1, 2, 4 ], [ 1, 2 ] ]
```
Comment (wilfwilson) [inline doc/conglatt.xml]: This will presumably change.

File: doc/conglatt.xml [hunk 4]
```diff
@@ -91,17 +91,22 @@ gap> Length(poset);
...
+ of size 3, degree 2 with 2 generators>>
 gap> LatticeOfLeftCongruences(S);
-[ [  ], [ 1, 3 ], [ 1 ] ]
+<poset of 3 congruences over <regular transformation monoid 
+ of size 3, degree 2 with 2 generators>>
 gap> LatticeOfRightCongruences(S);
+<poset of 5 congruences over <regular transformation monoid 
+ of size 3, degree 2 with 2 generators>>
+gap> OutNeighbours(LatticeOfRightCongruences(S));
 [ [  ], [ 1 ], [ 1 ], [ 1 ], [ 1, 2, 3, 4 ] ]
```
Comment (wilfwilson) [inline doc/conglatt.xml]: This will change

File: gap/congruences/conglatt.gi [hunk 5]
```diff
@@ -150,16 +140,16 @@ InstallMethod(JoinSemilatticeOfCongruences,
...
   # Trivial case
   if Size(poset) = 0 then
     return poset;
   fi;
 
   # Extract the info
-  children := OutNeighboursMutableCopy(poset!.po);
-  parents := InNeighboursMutableCopy(poset!.po);
+  children := OutNeighboursMutableCopy(poset);
+  parents := InNeighboursMutableCopy(poset);
```
Comment (wilfwilson) [inline gap/congruences/conglatt.gi]: swap this

File: gap/congruences/conglatt.gi [hunk 6]
```diff
@@ -220,12 +210,12 @@ function(poset, join_func)
   SEMIGROUPS.OptionsRec(S).report := report;
 
   # We are done: make the object and return
-  po := Digraph(children);
-  SetInNeighbours(po, parents);
-  poset := Objectify(NewType(FamilyObj(children), IsCongruencePoset),
-                     rec(po := po));
+  poset := Digraph(children);
```
Comment (wilfwilson) [inline gap/congruences/conglatt.gi]: poset := Digraph(parents);

File: gap/congruences/conglatt.gi [hunk 7]
```diff
@@ -220,12 +210,12 @@ function(poset, join_func)
...
+  SetInNeighbours(poset, parents);
```
Comment (wilfwilson) [inline gap/congruences/conglatt.gi]: SetInNeighbours(poset, children);

File: gap/congruences/conglatt.gi [hunk 8]
```diff
@@ -239,11 +229,11 @@ function(coll, join_func)
...
 
 SEMIGROUPS.AddTrivialCongruence := function(poset, cong_func)
-  local S, children, parents, congs, nrcongs, i, po;
+  local S, children, parents, congs, nrcongs, i;
   # Extract the info
   S := UnderlyingSemigroupOfCongruencePoset(poset);
-  children := OutNeighboursMutableCopy(poset!.po);
-  parents := InNeighboursMutableCopy(poset!.po);
+  children := OutNeighboursMutableCopy(poset);
+  parents := InNeighboursMutableCopy(poset);
```
Comment (wilfwilson) [inline gap/congruences/conglatt.gi]: swap

File: gap/congruences/conglatt.gi [hunk 9]
```diff
@@ -256,20 +246,20 @@ SEMIGROUPS.AddTrivialCongruence := function(poset, cong_func)
   od;
 
   # Make the object and return
-  po := Digraph(children);
-  SetInNeighbours(po, parents);
-  poset := Objectify(NewType(FamilyObj(children), IsCongruencePoset),
-                     rec(po := po));
+  poset := Digraph(children);
```
Comment (wilfwilson) [inline gap/congruences/conglatt.gi]: same swap

File: gap/congruences/conglatt.gi [hunk 10]
```diff
@@ -295,14 +285,14 @@ function(coll)
   od;
 
   # We are done: make the object and return
-  po := Digraph(children);
-  SetInNeighbours(po, parents);
-  poset := Objectify(NewType(FamilyObj(children), IsCongruencePoset),
-                     rec(po := po));
+  poset := Digraph(children);
```
Comment (wilfwilson) [inline gap/congruences/conglatt.gi]: same swap

File: gap/congruences/conglatt.gi [hunk 11]
```diff
@@ -530,10 +521,10 @@ function(poset, opts)
       fi;
     od;
   else
-    symbols := List([1 .. Length(poset)], String);
+    symbols := List([1 .. Length(out_nbs)], String);
   fi;
 
-  rel := List([1 .. Length(poset)], x -> Filtered(poset[x], y -> x <> y));
+  rel := List([1 .. Length(out_nbs)], x -> Filtered(out_nbs[x], y -> x <> y));
```
Comment (wilfwilson) [inline gap/congruences/conglatt.gi]: Change to `rel := InNeighbours(DigraphReflexiveTransitiveReduction(poset));`

The remove the line `j := Difference(rel[i], Union(rel{rel[i]}));`.

File: gap/congruences/conglatt.gi [hunk 12]
```diff
@@ -510,14 +500,15 @@ InstallMethod(DotString,
 "for a congruence poset and a record",
 [IsCongruencePoset, IsRecord],
 function(poset, opts)
-  local congs, S, symbols, i, nr, rel, str, j, k;
+  local out_nbs, congs, S, symbols, i, nr, rel, str, j, k;
+  out_nbs := OutNeighbours(poset);
```
Comment (wilfwilson) [inline gap/congruences/conglatt.gi]: Dont need this line. Throughout this method, change `Length(out_nbs)` and `Length(rel)` to `Size(poset)`.

**Comment by mtorpey** [comment]
Added changes to reverse the ordering.

I still need to add loops.

**Comment by mtorpey** [comment]
This is done!

**Pull request merged**